clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java

```diff
@@ -188,7 +194,14 @@ public CompletableFuture<Map<TopicPartition, OffsetAndTimestamp>> fetchOffsets(
      * this function (ex. {@link org.apache.kafka.common.errors.TopicAuthorizationException})
      */
     public void resetPositionsIfNeeded() {
-        Map<TopicPartition, Long> offsetResetTimestamps = offsetFetcherUtils.getOffsetResetTimestamp();
+        Map<TopicPartition, Long> offsetResetTimestamps;
+
+        try {
+            offsetResetTimestamps = offsetFetcherUtils.getOffsetResetTimestamp();
```

Comment:
In `offsetFetcherUtils.getOffsetResetTimestamp()`, if there is no offset reset policy, currently we just ignore it. It seems that we should throw an exception in that case?

Reply:
There is a default offset reset strategy which is configured via `ConsumerConfig`, passed to the `KafkaConsumer`, and then to the `SubscriptionState`, right? Why wouldn't we fall back to that if there's no overridden offset reset strategy for a particular partition?

Reply:
[auto.offset.reset](https://kafka.apache.org/documentation/#consumerconfigs_auto.offset.reset) has 3 options: `latest`, `earliest`, `none`. It's possible for a user to choose `none`, which means "throw exception to the consumer if no previous offset is found for the consumer's group". But in that case, `offsetFetcherUtils.getOffsetResetTimestamp` just ignores that partition and won't reset it forever. We need to throw an exception so that the user knows.

It's a bit of weird configuration since it means a consumer can't really consume for the very first time with this option. However, I tested it out with the existing consumer and it does throw an exception.
```
bin/kafka-console-consumer.sh --consumer-property auto.offset.reset=none --bootstrap-server localhost:9092 --topic test
[2023-10-20 15:30:07,088] ERROR Error processing message, terminating consumer process:  (kafka.tools.ConsoleConsumer$)
org.apache.kafka.clients.consumer.NoOffsetForPartitionException: Undefined offset with no reset policy for partitions: [test-0]
	at org.apache.kafka.clients.consumer.internals.SubscriptionState.resetInitializingPositions(SubscriptionState.java:711)
	at org.apache.kafka.clients.consumer.KafkaConsumer.updateFetchPositions(KafkaConsumer.java:2459)
	at org.apache.kafka.clients.consumer.KafkaConsumer.updateAssignmentMetadataIfNeeded(KafkaConsumer.java:1243)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1198)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1178)
	at kafka.tools.ConsoleConsumer$ConsumerWrapper.receive(ConsoleConsumer.scala:473)
	at kafka.tools.ConsoleConsumer$.process(ConsoleConsumer.scala:103)
	at kafka.tools.ConsoleConsumer$.run(ConsoleConsumer.scala:77)
	at kafka.tools.ConsoleConsumer$.main(ConsoleConsumer.scala:54)
	at kafka.tools.ConsoleConsumer.main(ConsoleConsumer.scala)
```

Reply:
I changed the `return null` to `throw new NoOffsetForPartitionException(partition)`. I re-ran the tests and nothing failed, so there seems to be another gap in the tests.